src/core/net/tcp6.hpp

```diff
@@ -329,7 +332,22 @@ class Tcp : public InstanceLocator, private NonCopyable
          */
         Error Deinitialize(void);
 
+        /**
+         * Converts a reference to a struct tcpcb to a reference to its
+         * enclosing Endpoint.
+         */
+        static Endpoint &FromTcb(struct tcpcb &aTcb)
+        {
+            uintptr_t ptr = reinterpret_cast<uintptr_t>(&aTcb);
+            ptr -= offsetof(Endpoint, mTcb);
```

Comment:
Is there a way to avoid using `offsetof()`?

I would avice avoid using it in general. It requires PoD types and can lead to some tricky bugs/issue (some toolchains have weird/unexpected behaviors).